Added fast correspondence estimation using a lookup table (40-60 times faster than using a k-d tree).

carlosmccosta opened the pull request:
Added fast correspondence estimation using a lookup table that precomputes the closest point between each voxel grid cell and a reference point cloud.

This approach can be 40 to 60 times faster than using a k-d tree to retrieve the closest point (simple memory access vs tree search) and when used with ICP it can make point cloud registration 5 to 9 times faster.

For small volumes or for 2d point cloud registration, a lookup table can provide a significant registration speedup at the cost of higher memory usage (in relation to a k-d tree).

For real-time applications, such as 3 or 6 DoF robot self-localization, in which the sensor data is continuously registered with a static / dynamic map, fast registration and low CPU load is typically more important then memory usage. As such, this approach might be more suitable than k-d trees.

Currently, the lookup table relies on a k-d tree to initialize each voxel grid cell.
For small volumes this usually takes less than a second, but as the volume / cell resolution increases the initialization can take some time.

In the future, i'm planning to use the distance transform approach to initialize the lookup table, given that it can be much faster.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+          * @return True if the lookup table was initialized successfully.
+          */
+        virtual bool
+        initLookupTable (typename pcl::search::Search<PointT>::ConstPtr tree);
```
Why is this method virtual?

Self: Does this need a reference?

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+          , number_of_queries_on_search_tree_ (0) {}
+
+        /** \brief Empty destructor. */
+        virtual ~CorrespondenceLookupTable () {}
```
Self: This destructor is only justifiable if there's a good reason to have virtual methods.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+          * @return True if the computed bounds are valid.
+          */
+        virtual bool
+        computeLookupTableBounds (const pcl::PointCloud<PointT>& pointcloud);
```
Why is this method virtual?

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+        computeSquaredDistance (const Eigen::Vector3f& a, const Eigen::Vector3f& b)
+        {
+          return ((a - b).squaredNorm ());
+        }
```
This is not worth it. Please delete.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+          float dy = a.y - b.y;
+          float dz = a.z - b.z;
+          return (dx * dx + dy * dy + dz * dz);
+        }
```
Delete the method and just use `(a.getVector3fMap () - b.getVector3fMap ()).squaredNorm ()`. This is not worth to have in the public interface of the class.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+          float dy = query_point.y - centroid (1);
+          float dz = query_point.z - centroid (2);
+          return (dx * dx + dy * dy + dz * dz);
+        }
```
Delete the method and just use `(query_point.getVector3fMap () - centroid).squaredNorm ()`.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+            return (true);
+          }
+          return (false);
+        }
```
Is this really necessary? It's a method to do bound checks on a vector... This can be checked in advance if really needed. Most of the times don't always need to check for the validity of the index.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+         * @return True if it was found a valid correspondence.
+         */
+        bool
+        getCorrespondence (const PointT& query_point, double maximum_correspondence_distance_squared, pcl::registration::CorrespondenceLookupTableCell& correspondance);
```
You need to provide a const alternative to this as well.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+        getCorrespondence (size_t correspondence_index)
+        {
+          return (lookup_table_[correspondence_index]);
+        }
```
You need to provide a const alternative to this method

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+        getCorrespondence (size_t x, size_t y, size_t z)
+        {
+          return (lookup_table_[x + y * number_cells_x_ + z * number_cells_xy_slice_]);
+        }
```
Const alternative

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+        size_t number_of_queries_on_lookup_table_;
+
+        /** \brief Number of queries performed on the search tree (because they were outside the lookup table bounds). */
+        size_t number_of_queries_on_search_tree_;
```
Why do you keep track of both of the members above?

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+        using CorrespondenceEstimationBase<PointSource, PointTarget, Scalar>::target_;
+        using CorrespondenceEstimationBase<PointSource, PointTarget, Scalar>::tree_;
+        using CorrespondenceEstimationBase<PointSource, PointTarget, Scalar>::tree_reciprocal_;
+        using PCLBase<PointSource>::deinitCompute;
```
All this lines with using should not be in public scope.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+        typedef typename PointCloudTarget::Ptr PointCloudTargetPtr;
+        typedef typename PointCloudTarget::ConstPtr PointCloudTargetConstPtr;
+
+        typedef typename KdTree::PointRepresentationConstPtr PointRepresentationConstPtr;
```
You don't need to redefine all of these type again. CorrespondenceEstimationBase already has them. Use the keyword `using` instead.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+
+        /** \brief Get the source lookup table. */
+        inline CorrespondenceLookupTable<PointSource>&
+        getSourceCorrespondencesLookupTable () { return (source_correspondences_lookup_table_); }
```
Do you really want to allow this to be changed from outside the class? If yes, you need to provide a const alternative.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+
+        /** \brief Get the target lookup table. */
+        inline CorrespondenceLookupTable<PointTarget>&
+        getTargetCorrespondencesLookupTable () { return (target_correspondences_lookup_table_); }
```
Do you really want to allow this to be changed from outside the class? If yes, you need to provide a const alternative.

SergioRAgostinho commented on registration/include/pcl/registration/correspondence_estimation_lookup_table.h:
```diff
@@ -0,0 +1,517 @@
+        {
+          CorrespondenceEstimationBase<PointSource, PointTarget, Scalar>::setSearchMethodSource (tree, force_no_recompute);
+          //if (!force_no_recompute)
+          //  source_correspondences_lookup_table_.initLookupTable(tree);
```
Delete these comments

SergioRAgostinho commented on registration/include/pcl/registration/impl/correspondence_estimation_lookup_table.hpp:
```diff
@@ -0,0 +1,324 @@
+  Eigen::Vector4f min_pt;
+  Eigen::Vector4f max_pt;
+  pcl::getMinMax3D (pointcloud, min_pt, max_pt);
+  float half_cell_resolution = cell_resolution_ * 0.5;
```
Add const qualifier and 0.5f

taketwo commented:
BTW, are there any related publications?

carlosmccosta commented:
Good afternoon :)
Thank you for the feedback :)
I am occupied with several research projects right now, and as such i will only have time to look into this PR in the next month.

This PR was used in the following journal article that is under review:
Héber Sobreira, Carlos M. Costa, Ivo Sousa, Luís Rocha, José Lima, Paulo C.M.A. Farias, Paulo Costa, and A. Paulo Moreira, "Map-Matching Algorithms for Robot Self-Localization: A comparison between Perfect Match, Iterative Closest Point and Normal Distributions Transform", Journal of Intelligent & Robotic Systems (JINT), 2017

Have a nice day :)

carlosmccosta commented:
In the master-all-pr branch I already implemented the Meijster distance transform for initializing 2D LUTs and added some unit tests:
https://github.com/carlosmccosta/pcl
I will extend it to 3D in the next month and merge it with this PR.

Have a nice day :)

taketwo commented:
Thanks for the pointers.

Regarding merging with this PR... I'd rather prefer separate PRs, otherwise reviewing becomes unmanageable.

carlosmccosta commented:
Ok.
I will address the suggestions given and add unit tests to this PR.
Then I will create another PR for the Meijster distance transform extension.

Have a nice day :)
